Scrollable target list + support for multiple config files

feinom opened the pull request:
If the number of targets exceeded the height of the terminal (number of lines), you couldn't see all the available targets. I have implemented scrolling in the target list to fix this. To get the best experience, you need the shutil.get_terminal_size() function, which was implemented in Python 3.3.

I have also added a simple argument parser to support multiple configuration files, in case there is a need for different menus for servers, switches or other stuff.

Commit:
Implemented scrolling in the list of targets to support more targets.
The height of the terminal is calculated by shutil.get_terminal_size(), which was implemented in Python 3.3. If this function isn't available, a fixed height is set instead.

Commit:
Added argument parser to support multiple config files. The argument is optional, and the default config name is config.json.

Mike724 commented:
Strangely `shutil.get_terminal_size` doesn't seem to work on macOS Sierra (Python 3.5.2), however `tput lines` does work.

Here's the code I used:
```
def get_terminal_height():
    """Returns the terminal's height"""
    height = os.popen('tput lines', 'r').readline()
    return int(height)
```

Assuming this also works on Linux, it might be a better choice (portability wise) since we already depend on `tput` for clearing the screen and moving the cursor.

mmeyer724 commented on sshmenu/sshmenu.py:
```diff
@@ -1,20 +1,27 @@
 
 def main():
+    # Check arguments
+    parser = argparse.ArgumentParser(prog='sshmenu', description='A simple tool for connecting to remote hosts via ssh.')
```
I was thinking "use clint!" but clint's argument parsing is... lacking.

Anyways, I think the description can be more clear. I like `A convenient tool for bookmarking hosts and connecting to them via ssh.`, although it's somewhat wordy.

mmeyer724 commented on sshmenu/sshmenu.py:
```diff
@@ -1,20 +1,27 @@
 def main():
+    # Check arguments
+    parser = argparse.ArgumentParser(prog='sshmenu', description='A simple tool for connecting to remote hosts via ssh.')
+    parser.add_argument('-c', '--configname', default='config.json', help='Specify an alternative configuration name.')
```
I think we should make the default value `'config'` and simply append `'.json'` to whatever value comes back. The file extension makes me feel like I can pass in an absolute path, which definitely isn't the case with clint.

mmeyer724 commented on sshmenu/sshmenu.py:
```diff
@@ -45,16 +61,23 @@ def display_menu(targets):
 
-    # Clear screen, give instructions
+        # Check line length
+        desc = '%2d ' % (index) + target['host'].ljust(longest_host) + ' | ' + target['friendly']
```
I think this should be split into a separate function, like `def get_target_desc(target):` in order to avoid this line being duplicated further down.

Another alternative is to store the value in target['desc'] and reference that value further down.

Mike724 commented:
I added a few small suggestions, nothing major. After changing out `shutil.get_terminal_size` for `tput lines` everything seems to work as expected.

This is definitely a much needed feature, thanks for sending this in.

feinom commented:
Thanks! I can make the changes you pointed out tonight. They make sense to me :-)

I wasn't able to get 'tput lines'  to report the correct number of lines when loading sshmenu as a module. It would report '24' no matter what the height actually was. We could perhaps use 'tput' on OS X, and 'shutil' on Linux?

Mike724 commented:
Hmm, which terminal emulator are you using, and what is command you're running to test changes? I use `python -m ssshmenu` while in the top level directory of the repository for development. 

`stty size` is another option, if a portable solution doesn't exist I'm ok with per-os logic.

feinom commented:
I played around some more with tput, and I managed to figure out the problem. The problem surfaces with this code:

```
tput = Popen(["tput", "lines"], stdout=PIPE, stderr=PIPE)
height, stderr = tput.communicate()
```
In the above case, `height` will always get `24` as it's value.

However, if I remove `stderr=PIPE` from the code, it returns the correct terminal height. Go figure :-) So this code works for me:

```
tput = Popen(["tput", "lines"], stdout=PIPE)
height, stderr = tput.communicate()
```

I'll switch from shutil.get_terminal_size() to using tput now, but if you know why `stderr=PIPE` causes this behavior, I'd like to know!

Commit:
Changed from shutil.get_terminal_size() to calling "tput lines".
Changed app description in the help output.
".json" is now appended to config names passed in through the "-c" switch.
Target "desc" lines are now only generated in one place, and is stored within the target for later use.

feinom commented:
By the way, I would also like to add support for the Page Up and Page Down keys, but there seems to be a bug in the readchar library. It looks like they've fixed it, but they haven't made any new releases in a while., so I don't think I'll get it into this pull request. Link to the relevant commit: https://github.com/magmax/python-readchar/commit/7437eb2a4968ecca9a0b07de5a6d6ef165a99e44

Commit:
Improve detection of when the terminal window has been resized.

Mike724 commented:
Eh, I'd rather stick with tput, considering sshmenu already requires it.

Commit:
There was a problem with the resizing logic in the last commit. Fixed it now, so it should be behaving correctly.

feinom commented:
Yeah, I agree. I think I'm done adding stuff for now, but let me know if you spot anything that can be improved. It's a fun project to work on :-)

Mike724 commented:
No problem! I'm working on a few modifications myself (invisible cursor while in the menu, fully clearing the screen on resize, and potentially truncating long lines with an ellipsis).

Once that's done I'll release 0.0.4. 👍

feinom commented:
Sounds great! Looking forward to testing it :-)
